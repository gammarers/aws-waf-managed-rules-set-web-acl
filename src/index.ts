import * as waf from 'aws-cdk-lib/aws-wafv2';
import { Construct } from 'constructs';

export interface WafManagedRulesSetWebACLProps {
  readonly name?: string;
  readonly scope: ScopeType;
}

export enum ScopeType {
  GLOBAL = 'Global',
  REGIONAL = 'Regional',
}

// export enum ServerType {
//   LINUX = 'Linux',
//   WINDOWS = 'Windows',
// }

export class WafManagedRulesSetWebACL extends waf.CfnWebACL {
  constructor(scope: Construct, id: string, props: WafManagedRulesSetWebACLProps) {
    super(scope, id, {
      name: props.name,
      description: '',
      scope: ((): string => {
        switch (props.scope) {
          case ScopeType.GLOBAL:
            return 'CLOUDFRONT';
          case ScopeType.REGIONAL:
            return 'REGIONAL';
        }
      })(),
      defaultAction: {
        allow: {},
      },
      rules: ((): waf.CfnWebACL.RuleProperty[] => {
        //const rules: waf.CfnWebACL.RuleProperty[] = [
        return [
          {
            name: 'aws-managed-rules-common-rule-set',
            priority: 0,
            overrideAction: {
              count: {},
            },
            statement: {
              managedRuleGroupStatement: {
                vendorName: 'AWS',
                name: 'AWSManagedRulesCommonRuleSet',
                excludedRules: [],
              },
            },
            visibilityConfig: {
              sampledRequestsEnabled: true,
              cloudWatchMetricsEnabled: true,
              metricName: 'AWSManagedRulesCommonRuleSet',
            },
          },
          {
            name: 'aws-managed-rules-known-bad-inputs-rule-set',
            priority: 1,
            overrideAction: {
              count: {},
            },
            statement: {
              managedRuleGroupStatement: {
                vendorName: 'AWS',
                name: 'AWSManagedRulesKnownBadInputsRuleSet',
                excludedRules: [],
              },
            },
            visibilityConfig: {
              sampledRequestsEnabled: true,
              cloudWatchMetricsEnabled: true,
              metricName: 'AWSManagedRulesKnownBadInputsRuleSet',
            },
          },
          {
            name: 'aws-managed-rules-amazon-ip-reputation-rule-set',
            priority: 2,
            overrideAction: {
              count: {},
            },
            statement: {
              managedRuleGroupStatement: {
                vendorName: 'AWS',
                name: 'AWSManagedRulesAmazonIpReputationList',
                excludedRules: [],
              },
            },
            visibilityConfig: {
              sampledRequestsEnabled: true,
              cloudWatchMetricsEnabled: true,
              metricName: 'AWSManagedRulesAmazonIpReputationList',
            },
          },
          {
            name: 'aws-managed-rules-anonymous-ip-list',
            priority: 3,
            overrideAction: {
              count: {},
            },
            statement: {
              managedRuleGroupStatement: {
                vendorName: 'AWS',
                name: 'AWSManagedRulesAnonymousIpList',
                excludedRules: [],
              },
            },
            visibilityConfig: {
              sampledRequestsEnabled: true,
              cloudWatchMetricsEnabled: true,
              metricName: 'AWSManagedRulesAnonymousIpList',
            },
          },
          {
            name: 'aws-managed-rules-sqli-rule-set',
            priority: 4,
            overrideAction: {
              count: {},
            },
            statement: {
              managedRuleGroupStatement: {
                vendorName: 'AWS',
                name: 'AWSManagedRulesSQLiRuleSet',
                excludedRules: [],
              },
            },
            visibilityConfig: {
              sampledRequestsEnabled: true,
              cloudWatchMetricsEnabled: true,
              metricName: 'AWSManagedRulesSQLiRuleSet',
            },
          },
          // only server type is linux
          {
            name: 'aws-managed-rules-linux-rule-set',
            priority: 5,
            overrideAction: {
              count: {},
            },
            statement: {
              managedRuleGroupStatement: {
                vendorName: 'AWS',
                name: 'AWSManagedRulesLinuxRuleSet',
                excludedRules: [],
              },
            },
            visibilityConfig: {
              sampledRequestsEnabled: true,
              cloudWatchMetricsEnabled: true,
              metricName: 'AWSManagedRulesLinuxRuleSet',
            },
          },
          {
            name: 'aws-managed-rules-unix-rule-set',
            priority: 6,
            overrideAction: {
              count: {},
            },
            statement: {
              managedRuleGroupStatement: {
                vendorName: 'AWS',
                name: 'AWSManagedRulesUnixRuleSet',
                excludedRules: [],
              },
            },
            visibilityConfig: {
              sampledRequestsEnabled: true,
              cloudWatchMetricsEnabled: true,
              metricName: 'AWSManagedRulesUnixRuleSet',
            },
          },
        ];
        //return rules;
      })(),
      customResponseBodies: undefined,
      visibilityConfig: {
        sampledRequestsEnabled: true,
        cloudWatchMetricsEnabled: true,
        metricName: 'WafManagedRulesSetWebACL',
      },
    });
  }
}