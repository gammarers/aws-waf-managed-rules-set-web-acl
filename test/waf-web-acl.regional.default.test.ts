import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as waf from 'aws-cdk-lib/aws-wafv2';
import { ScopeType, WafManagedRulesSetWebACL } from '../src';

describe('Web Acl rule default group testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  const acl = new WafManagedRulesSetWebACL(stack, 'WafManagedRulesSetWebACL', {
    scope: ScopeType.GLOBAL,
  });

  it('Is Waf Web ACL', () => {
    expect(acl).toBeInstanceOf(waf.CfnWebACL);
  });

  const template = Template.fromStack(stack);


  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});