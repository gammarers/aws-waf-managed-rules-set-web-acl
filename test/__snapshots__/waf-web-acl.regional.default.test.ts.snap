// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web Acl rule default group testing Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "WafManagedRulesSetWebACL": Object {
      "Properties": Object {
        "DefaultAction": Object {
          "Allow": Object {},
        },
        "Description": "",
        "Rules": Array [
          Object {
            "Name": "aws-managed-rules-common-rule-set",
            "OverrideAction": Object {
              "Count": Object {},
            },
            "Priority": 0,
            "Statement": Object {
              "ManagedRuleGroupStatement": Object {
                "ExcludedRules": Array [],
                "Name": "AWSManagedRulesCommonRuleSet",
                "VendorName": "AWS",
              },
            },
            "VisibilityConfig": Object {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "AWSManagedRulesCommonRuleSet",
              "SampledRequestsEnabled": true,
            },
          },
          Object {
            "Name": "aws-managed-rules-known-bad-inputs-rule-set",
            "OverrideAction": Object {
              "Count": Object {},
            },
            "Priority": 1,
            "Statement": Object {
              "ManagedRuleGroupStatement": Object {
                "ExcludedRules": Array [],
                "Name": "AWSManagedRulesKnownBadInputsRuleSet",
                "VendorName": "AWS",
              },
            },
            "VisibilityConfig": Object {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "AWSManagedRulesKnownBadInputsRuleSet",
              "SampledRequestsEnabled": true,
            },
          },
          Object {
            "Name": "aws-managed-rules-amazon-ip-reputation-rule-set",
            "OverrideAction": Object {
              "Count": Object {},
            },
            "Priority": 2,
            "Statement": Object {
              "ManagedRuleGroupStatement": Object {
                "ExcludedRules": Array [],
                "Name": "AWSManagedRulesAmazonIpReputationList",
                "VendorName": "AWS",
              },
            },
            "VisibilityConfig": Object {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "AWSManagedRulesAmazonIpReputationList",
              "SampledRequestsEnabled": true,
            },
          },
          Object {
            "Name": "aws-managed-rules-anonymous-ip-list",
            "OverrideAction": Object {
              "Count": Object {},
            },
            "Priority": 3,
            "Statement": Object {
              "ManagedRuleGroupStatement": Object {
                "ExcludedRules": Array [],
                "Name": "AWSManagedRulesAnonymousIpList",
                "VendorName": "AWS",
              },
            },
            "VisibilityConfig": Object {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "AWSManagedRulesAnonymousIpList",
              "SampledRequestsEnabled": true,
            },
          },
          Object {
            "Name": "aws-managed-rules-sqli-rule-set",
            "OverrideAction": Object {
              "Count": Object {},
            },
            "Priority": 4,
            "Statement": Object {
              "ManagedRuleGroupStatement": Object {
                "ExcludedRules": Array [],
                "Name": "AWSManagedRulesSQLiRuleSet",
                "VendorName": "AWS",
              },
            },
            "VisibilityConfig": Object {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "AWSManagedRulesSQLiRuleSet",
              "SampledRequestsEnabled": true,
            },
          },
          Object {
            "Name": "aws-managed-rules-linux-rule-set",
            "OverrideAction": Object {
              "Count": Object {},
            },
            "Priority": 5,
            "Statement": Object {
              "ManagedRuleGroupStatement": Object {
                "ExcludedRules": Array [],
                "Name": "AWSManagedRulesLinuxRuleSet",
                "VendorName": "AWS",
              },
            },
            "VisibilityConfig": Object {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "AWSManagedRulesLinuxRuleSet",
              "SampledRequestsEnabled": true,
            },
          },
          Object {
            "Name": "aws-managed-rules-unix-rule-set",
            "OverrideAction": Object {
              "Count": Object {},
            },
            "Priority": 6,
            "Statement": Object {
              "ManagedRuleGroupStatement": Object {
                "ExcludedRules": Array [],
                "Name": "AWSManagedRulesUnixRuleSet",
                "VendorName": "AWS",
              },
            },
            "VisibilityConfig": Object {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "AWSManagedRulesUnixRuleSet",
              "SampledRequestsEnabled": true,
            },
          },
        ],
        "Scope": "CLOUDFRONT",
        "VisibilityConfig": Object {
          "CloudWatchMetricsEnabled": true,
          "MetricName": "WafManagedRulesSetWebACL",
          "SampledRequestsEnabled": true,
        },
      },
      "Type": "AWS::WAFv2::WebACL",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
